Added custom mac addresses + dhcp renew, longer tcp ack retransmit timeout, improved error handling.

BlakeGarner opened the pull request:
Added support for:
- begin(byte *mac) {...} lets you specify your own MAC address [begin() is still supported]
- Fix to not send "Content-Type" in GET requests.
- Close connection if data size error, so it can restart gracefully.
- Added maintain() method that should be called periodically to refresh DHCP settings.
- Retransmission Time (ie time to retransmit if receive no TCP ACK) set to 4s, default too short.

BlakeGarner committed:
Added support for:
+ begin(byte *mac) {...} lets you specify your own MAC address [begin() is still supported]
+ Fix to not send "Content-Type" in GET requests.
+ Close connection if data size error, so it can restart gracefully.
+ Added maintain() method that should be called periodically to refresh DHCP settings.

Signed-off-by: BlakeGarner <blake.garner@gmail.com>

wolfeidau commented:
Hey what was the reason for the custom mac address? 

Is this a common thing to need to override on arduino?

BlakeGarner commented:
Hey,
The W5100-based Ethernet shield on Arduino does not have a hardware-allocated MAC address - it must be set in code. The arduino-ninja-blocks library had a single mac hard-coded in - which is fine for prototyping, but problematic for long-term use, particularly if two or more devices were going to be deployed on the same LAN utilising this library.

wolfeidau commented:
Ahh just reviewing the examples on the Arduino site.

http://arduino.cc/en/Tutorial/WebClient

Has a note in the code associated with the mac address.

``` c
// Enter a MAC address for your controller below.
// Newer Ethernet shields have a MAC address printed on a sticker on the shield
byte mac[] = { 0xDE, 0xAD, 0xBE, 0xEF, 0xFE, 0xED };
```

Given this is labeled a library we should certainly enable this to be overridden.

BlakeGarner committed:
Fixed TCP fragmentation, string optimisations.

Previously single HTTP GET or POST messages were broken up across
multiple TCP packets. While okay, it did put additional load on the SPI
and Ethernet interfaces and could be fixed by preparing the entire
message ahead of time in a local buffer.
This did require some rearrangement of how httppost and ninjaMessage
were run, and a buffer (448 bytes).
The extra memory used by the buffer has mostly been reclaimed through
other optimisations:
- Fixed strings use program (flash) memory instead of SRAM.
- Build the JSON data on the fly within the packet buffer, rather than
in its own.
- Use the same buffer for reading and sending packets to/from the W5100,
as they can't occur at the same time.

Other Changes:
- strcat has a O(n) penalty to find the end of the string every time it
is run. Removed it in favour of strcpy and keeping track of where the
pointer is up to.
- Removed a few redundant or unused library or variable declarations.
- Added a few extra Serial.prints to help debug connection errors in the
future.
